File: csharp/src/DataStructures/MinHeap.cs
namespace DSA.DataStructures;

/// <summary>
/// Binary min-heap implementation.
/// Time: O(log n) insert/extract, O(1) peek
/// Space: O(n)
/// </summary>
public class MinHeap<T>
{
    private readonly List<T> _heap = new();
    private readonly IComparer<T> _comparer;

    public int Count => _heap.Count;
    public bool IsEmpty => _heap.Count == 0;

    public MinHeap(IComparer<T>? comparer = null)
    {
        _comparer = comparer ?? Comparer<T>.Default;
    }

    public void Insert(T value)
    {
        _heap.Add(value);
        SiftUp(_heap.Count - 1);
    }

    public T? ExtractMin()
    {
        if (_heap.Count == 0)
            return default;

        var min = _heap[0];
        var last = _heap[^1];
        _heap.RemoveAt(_heap.Count - 1);

        if (_heap.Count > 0)
        {
            _heap[0] = last;
            SiftDown(0);
        }

        return min;
    }

    public T? Peek() => _heap.Count == 0 ? default : _heap[0];

    public void Clear() => _heap.Clear();

    public T[] ToArray() => _heap.ToArray();

    private void SiftUp(int index)
    {
        while (index > 0)
        {
            var parentIndex = (index - 1) / 2;
            if (_comparer.Compare(_heap[index], _heap[parentIndex]) >= 0)
                break;

            Swap(index, parentIndex);
            index = parentIndex;
        }
    }

    private void SiftDown(int index)
    {
        while (true)
        {
            var leftChild = 2 * index + 1;
            var rightChild = 2 * index + 2;
            var smallest = index;

            if (leftChild < _heap.Count && _comparer.Compare(_heap[leftChild], _heap[smallest]) < 0)
                smallest = leftChild;

            if (rightChild < _heap.Count && _comparer.Compare(_heap[rightChild], _heap[smallest]) < 0)
                smallest = rightChild;

            if (smallest == index)
                break;

            Swap(index, smallest);
            index = smallest;
        }
    }

    private void Swap(int i, int j)
    {
        (_heap[i], _heap[j]) = (_heap[j], _heap[i]);
    }
}

File: php/src/DataStructures/MinHeap.php
<?php

declare(strict_types=1);

namespace DSA\DataStructures;

/**
 * Binary min-heap implementation.
 * Time: O(log n) insert/extract, O(1) peek
 * Space: O(n)
 *
 * @template T
 */
class MinHeap
{
    private array $heap = [];
    /** @var callable */
    private $comparator;

    public function __construct(?callable $comparator = null)
    {
        $this->comparator = $comparator ?? fn($a, $b) => $a <=> $b;
    }

    public function insert(mixed $value): void
    {
        $this->heap[] = $value;
        $this->siftUp(count($this->heap) - 1);
    }

    public function extractMin(): mixed
    {
        if (empty($this->heap)) {
            return null;
        }

        $min = $this->heap[0];
        $last = array_pop($this->heap);

        if (!empty($this->heap)) {
            $this->heap[0] = $last;
            $this->siftDown(0);
        }

        return $min;
    }

    public function peek(): mixed
    {
        return $this->heap[0] ?? null;
    }

    public function isEmpty(): bool
    {
        return empty($this->heap);
    }

    public function size(): int
    {
        return count($this->heap);
    }

    public function clear(): void
    {
        $this->heap = [];
    }

    public function toArray(): array
    {
        return $this->heap;
    }

    private function siftUp(int $index): void
    {
        while ($index > 0) {
            $parentIndex = intdiv($index - 1, 2);
            if (($this->comparator)($this->heap[$index], $this->heap[$parentIndex]) >= 0) {
                break;
            }
            $this->swap($index, $parentIndex);
            $index = $parentIndex;
        }
    }

    private function siftDown(int $index): void
    {
        $length = count($this->heap);

        while (true) {
            $leftChild = 2 * $index + 1;
            $rightChild = 2 * $index + 2;
            $smallest = $index;

            if ($leftChild < $length &&
                ($this->comparator)($this->heap[$leftChild], $this->heap[$smallest]) < 0) {
                $smallest = $leftChild;
            }

            if ($rightChild < $length &&
                ($this->comparator)($this->heap[$rightChild], $this->heap[$smallest]) < 0) {
                $smallest = $rightChild;
            }

            if ($smallest === $index) {
                break;
            }

            $this->swap($index, $smallest);
            $index = $smallest;
        }
    }

    private function swap(int $i, int $j): void
    {
        [$this->heap[$i], $this->heap[$j]] = [$this->heap[$j], $this->heap[$i]];
    }
}
